File: solutions/clojure/082-linked-list-cycle.clj
;; Problem 82: Linked List Cycle (LeetCode 141)
;; Difficulty: Easy
;; Language: Clojure
;; 
(defn has-cycle? [head]
  (loop [slow head fast (when head (:next head))]
    (cond
      (or (nil? fast) (nil? (:next fast))) false
      (identical? slow fast) true
      :else (recur (:next slow) (:next (:next fast))))))

File: solutions/elixir/082-linked-list-cycle.ex
# Problem 82: Linked List Cycle (LeetCode 141)
# Difficulty: Easy
# Language: Elixir
# 
defmodule CycleDetect do
  def has_cycle?(nil), do: false
  def has_cycle?(head), do: detect(head, next(head))

  defp detect(_, nil), do: false
  defp detect(slow, fast) when slow == fast, do: true
  defp detect(slow, fast) do
    case next(fast) do
      nil -> false
      fast2 -> detect(next(slow), next(fast2))
    end
  end

  defp next(nil), do: nil
  defp next(%{next: n}), do: n
end
